rtengine/ahd_demosaic_RT.cc

```diff
@@ -0,0 +1,229 @@
+                    }
+                }
+
+            //  Build homogeneity maps from the CIELab images:
+
+            for (int row = top + 2; row < top + TS - 2 && row < height - 4; row++) {
+                int tr = row - top;
+
+                for (int col = left + 2, tc = 2; col < left + TS - 2 && col < width - 4; col++, tc++) {
+                    for (int d = 0; d < 2; d++) {
+                        lix = &lab[d][tr][tc];
+
+                        for (int i = 0; i < 4; i++) {
+                            ldiff[d][i] = std::fabs(lix[0][0] - lix[dir[i]][0]);
+                            abdiff[d][i] = SQR(lix[0][1] - lix[dir[i]][1])
+                                           + SQR(lix[0][2] - lix[dir[i]][2]);
+                        }
+                    }
+
+                    float leps = min(max(ldiff[0][0], ldiff[0][1]),
+                                     max(ldiff[1][2], ldiff[1][3]));
+                    float abeps = min(max(abdiff[0][0], abdiff[0][1]),
+                                      max(abdiff[1][2], abdiff[1][3]));
+
+                    for (int d = 0; d < 2; d++) {
+                        homo[d][tr][tc] = 0;
+                        for (int i = 0; i < 4; i++) {
+                            homo[d][tr][tc] += ((ldiff[d][i] <= leps) * (abdiff[d][i] <= abeps));
+                        }
+                    }
+                }
+            }
+
+            //  Combine the most homogenous pixels for the final result:
+            for (int row = top + 3; row < top + TS - 3 && row < height - 5; row++) {
+                int tr = row - top;
+
+                for (int col = left + 3, tc = 3; col < std::min(left + TS - 3, width - 5); col++, tc++) {
+                    uint16_t hm0 = 0, hm1 = 0;
+                    for (int i = tr - 1; i <= tr + 1; i++)
+                        for (int j = tc - 1; j <= tc + 1; j++) {
+                            hm0 += homo[0][i][j];
+                            hm1 += homo[1][i][j];
+                        }
+
+                    if (hm0 != hm1) {
+                        int dir = hm1 > hm0;
+                        red[row][col] = rgb[dir][tr][tc][0];
+                        green[row][col] = rgb[dir][tr][tc][1];
+                        blue[row][col] = rgb[dir][tr][tc][2];
+                    } else {
+                        red[row][col] = 0.5f * (rgb[0][tr][tc][0] + rgb[1][tr][tc][0]);
+                        green[row][col] = 0.5f * (rgb[0][tr][tc][1] + rgb[1][tr][tc][1]);
+                        blue[row][col] = 0.5f * (rgb[0][tr][tc][2] + rgb[1][tr][tc][2]);
+                    }
+                }
+            }
+
+            if(plistener) {
+                progresscounter++;
+
+                if(progresscounter % 32 == 0) {
+#ifdef _OPENMP
+                    #pragma omp critical (ahdprogress)
+#endif
+                    {
+                        progress += (double)32 * ((TS - 32) * (TS - 32)) / (height * width);
```

Comment:
Don't we have a `^2` function?

`(double)32` -> `32.0`.

Reply:
Done